### PR TITLE
 GTEST/STREAM: add generic datatype testing on send side

### PR DESCRIPTION
**Pull request by evgeny-leksikov**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6112/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by evgeny-leksikov** [comment]
@alex-mikheev pls review

File: test/gtest/common/test_helpers.cc [hunk 1]
```diff
@@ -208,4 +208,128 @@ data_type_desc_t::make(ucp_datatype_t datatype, const void *buf, size_t length,
...
+                    dt_state->count - (offset / sizeof(T)));
+    for (unsigned i = 0; i < count; ++i) {
+        p[i] = (offset / sizeof(T)) + i;
+    }
+    return count * sizeof(T);
+}
+
+template <typename T>
+ucs_status_t dt_unpack(void *state, size_t offset, const void *src,
+                                     size_t length)
```
Comment (alex-mikheev) [inline test/gtest/common/test_helpers.cc]: style: param align

File: test/gtest/common/test_helpers.cc [hunk 2]
```diff
@@ -208,4 +208,128 @@ data_type_desc_t::make(ucp_datatype_t datatype, const void *buf, size_t length,
...
+                           expected << " actual: " << actual << " offset: " <<
+                           offset << ".");
+        }
+    }
+    return UCS_OK;
+}
+
+static ucs_status_t dt_err_unpack(void *state, size_t offset, const void *src,
+                           size_t length)
+{
```
Comment (alex-mikheev) [inline test/gtest/common/test_helpers.cc]: style: param align

File: test/gtest/common/test_helpers.cc [hunk 3]
```diff
@@ -208,4 +208,128 @@ data_type_desc_t::make(ucp_datatype_t datatype, const void *buf, size_t length,
...
 }
 
+const uint32_t MAGIC    = 0xd7d7d7d7U;
+int dt_gen_start_count  = 0;
+int dt_gen_finish_count = 0;
+
+static void* dt_common_start(size_t count)
+{
+    dt_gen_state *dt_state = new dt_gen_state;
+    dt_state->count   = count;
```
Comment (alex-mikheev) [inline test/gtest/common/test_helpers.cc]: style: add empty line between code and decl

File: test/gtest/common/test_helpers.cc [hunk 4]
```diff
@@ -208,4 +208,128 @@ data_type_desc_t::make(ucp_datatype_t datatype, const void *buf, size_t length,
...
+static void* dt_common_start_unpack(void *context, void *buffer, size_t count)
+{
+    return dt_common_start(count);
+}
+
+template <typename T>
+size_t dt_packed_size(void *state)
+{
+    dt_gen_state *dt_state = (dt_gen_state*)state;
+    return dt_state->count * sizeof(T);
```
Comment (alex-mikheev) [inline test/gtest/common/test_helpers.cc]: style: add line between decl
 and return

File: test/gtest/common/test_helpers.cc [hunk 5]
```diff
@@ -208,4 +208,128 @@ data_type_desc_t::make(ucp_datatype_t datatype, const void *buf, size_t length,
...
+    EXPECT_GT(dt_gen_start_count, dt_gen_finish_count);
+    EXPECT_EQ(1, dt_state->started);
+    EXPECT_EQ(uint32_t(MAGIC), dt_state->magic);
+
+    return UCS_ERR_NO_MEMORY;
+}
+
+static void dt_common_finish(void *state)
+{
+    dt_gen_state *dt_state = (dt_gen_state*)state;
```
Comment (alex-mikheev) [inline test/gtest/common/test_helpers.cc]: style: add empty line

File: test/gtest/common/test_helpers.h [hunk 6]
```diff
@@ -31,6 +31,122 @@ extern "C" {
...
+/* Test output */
+#define UCS_TEST_MESSAGE \
+    ucs::detail::message_stream("INFO")
+
+
+/* Skip test */
+#define UCS_TEST_SKIP \
+    do { \
+        throw ucs::test_skip_exception(); \
+    } while(0)
```
Comment (alex-mikheev) [inline test/gtest/common/test_helpers.h]: style: add line

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6133/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6132/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6134/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
@alex-mikheev all comments are addressed.

**Comment by evgeny-leksikov** [comment]
@yosefe

**Comment by yosefe** [comment]
why adding generic type to common test_helpers? isn't this only for UCP?

**Comment by evgeny-leksikov** [comment]
it's under `ucp` name space

**Comment by yosefe** [comment]
so why not put it in gtest/ucp ?

**Comment by evgeny-leksikov** [comment]
because `data_type_desc_t` is placed in helpers

**Comment by yosefe** [comment]
if data_type_desc_t can be used only for UCP (since uct API does not define datatypes), it should also be moved to gtest/ucp

**Comment by evgeny-leksikov** [comment]
then, all `uct` namespace should be moved to gtest/uct?

**Comment by yosefe** [comment]
Yes (i guess you mean ucp?)

**Comment by evgeny-leksikov** [comment]
forgot... it was `ucs`, ok will move. into the `class ucp_test` or keep namespace?

**Comment by yosefe** [comment]
I would add a file for every class e.g separate file for datatype handling

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6138/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6141/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6142/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6143/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6145/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
@yosefe tests are passed

File: test/gtest/common/test_helpers.h [hunk 7]
```diff
@@ -31,6 +25,124 @@ extern "C" {
 #include <stdint.h>
 
 
+#ifndef UINT16_MAX
+#define UINT16_MAX (65535)
```
Comment (yosefe) [inline test/gtest/common/test_helpers.h]: why is this needed ? can use `#define __STDC_LIMIT_MACROS` before including `<stdint.h>`

File: test/gtest/common/test_helpers.h [hunk 8]
```diff
@@ -31,6 +25,124 @@ extern "C" {
...
+    } while(0)
+
+
+/* UCS error check */
+#define EXPECT_UCS_OK(_expr) \
+    do { \
+        ucs_status_t _status = (_expr); \
+        EXPECT_EQ(UCS_OK, _status) << "Error: " << ucs_status_string(_status); \
+    } while (0)
+
```
Comment (yosefe) [inline test/gtest/common/test_helpers.h]: code style: need 2 space lines between macros in header file

File: test/gtest/common/test_helpers.h (hunk 7 above)
Comment (evgeny-leksikov) [reply test/gtest/common/test_helpers.h]: `#define __STDC_LIMIT_MACROS` does not help, the problem is that we include internal C-headers directly from gtests. without this definition compilation is broken on ARM, If it's critical, i would prefer to play with this out of scope of this PR.

File: test/gtest/common/test_helpers.h (hunk 8 above)
Comment (evgeny-leksikov) [reply test/gtest/common/test_helpers.h]: Fixed

File: test/gtest/common/test_helpers.h (hunk 7 above)
Comment (yosefe) [reply test/gtest/common/test_helpers.h]: what is the problem to add this definition? we already do it in multiple tests (e.g ucm/malloc_hook.cc)

Comment (evgeny-leksikov) [reply test/gtest/common/test_helpers.h]: Adding in this file doesn't help, i did not try to do that on every compilation failure in every test.

Comment (yosefe) [reply test/gtest/common/test_helpers.h]: BTW, doesn't seem like UINT16_MAX is even used anywhere in gtest..
also, AFAIK the C++ way is using `std::numeric_limits<uint16_t>::max()`

Comment (evgeny-leksikov) [reply test/gtest/common/test_helpers.h]: Yes, but it's used in internal (non API) UC<X> headers included directly from gtests.

Comment (evgeny-leksikov) [reply test/gtest/common/test_helpers.h]: Probable right way is to include <stdint.h> in these library files, not in gtests

Comment (yosefe) [reply test/gtest/common/test_helpers.h]: IC. so need to either define __STDC_LIMIT_MACROS in all tests which include internal UCX headers and use these macros, or add it to gtest compilation command line CPPFLAGS (-D__STDC_LIMIT_MACROS)

Comment (evgeny-leksikov) [reply test/gtest/common/test_helpers.h]: Let's do it out of this PR. Really, this code was not even changed.

Comment (yosefe) [reply test/gtest/common/test_helpers.h]: ok

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6162/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6169/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6175/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6187/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6191/ for details (Mellanox internal link).
